pyadjoint/optimization/tao_solver.py

```diff
@@ -37,28 +37,35 @@ def __init__(self, X, *, comm=None):
             raise RuntimeError("PETSc not available")
 
         X = Enlist(X)
+        vecs, indices = self._get_vecs_indexes(X)
+        _, isis = PETSc.Vec().concatenate(vecs)
+        self._concatenate_vecs = vecs
+        self._isis_concatenated_vecs = [i for i in isis]
         if comm is None:
             comm = PETSc.COMM_WORLD
         if hasattr(comm, "tompi4py"):
             comm = comm.tompi4py()
-
-        indices = []
         n = 0
         N = 0
-        for x in X:
-            y = x._ad_copy()
-            y_a = np.zeros(y._ad_dim(), dtype=PETSc.ScalarType)
-            _, x_n = y._ad_assign_numpy(y, y_a, offset=0)
-            del y, y_a
-            indices.append((n, n + x_n))
-            n += x_n
+        for x, i in zip(X, indices):
+            indices.append((n, n + i))
+            n += i
             N += x._ad_dim()
-
+    
         self._comm = comm
         self._indices = tuple(indices)
         self._n = n
         self._N = N
 
+    def _get_vecs_indexes(self, X):
+        indices = []
+        vecs = []
+        for x in X:
+            vec = x._ad_vec_to_petsc()
+            indices.append(vec.getLocalSize())
```

Comment:
What do you think about `vec = OverloadedType._ad_vec_to_petsc()`? where
```
def _ad_vec_to_petsc(self):
        with self.dat.vec_ro as self_v:
            new_vec = self_v.copy()
        return new_vec
```

Reply:
Just `_ad_to_petsc`/`_ad_from_petsc`?

Reply:
Yes. It can be.